File: bindings/java/src/main/java/com/ainos/sdk/models/AinosRateLimitException.java
package com.ainos.sdk.models;

/**
 * Exception thrown when the daemon returns a rate limit error (HTTP 429).
 * <p>
 * Indicates that the client has exceeded the allowed request rate for
 * a particular API category. The {@link #getRetryAfterSeconds()} method
 * provides the suggested wait time before retrying.
 */
public class AinosRateLimitException extends AinosException {

    private static final long serialVersionUID = 1L;

    private final long retryAfterSeconds;

    public AinosRateLimitException(String message) {
        super(message);
        this.retryAfterSeconds = 1;
    }

    public AinosRateLimitException(String message, long retryAfterSeconds) {
        super(message);
        this.retryAfterSeconds = retryAfterSeconds;
    }

    public AinosRateLimitException(String message, long retryAfterSeconds, Throwable cause) {
        super(message, cause);
        this.retryAfterSeconds = retryAfterSeconds;
    }

    /**
     * Returns the number of seconds the client should wait before retrying.
     *
     * @return retry-after duration in seconds
     */
    public long getRetryAfterSeconds() {
        return retryAfterSeconds;
    }
}

File: bindings/csharp/AinosSdk/Models/AinosRateLimitException.cs
namespace AinosSdk.Models;

/// <summary>
/// Raised when the daemon returns a rate limit error (HTTP 429).
/// </summary>
public class AinosRateLimitException : AinosException
{
    /// <summary>
    /// Number of seconds to wait before retrying.
    /// </summary>
    public int RetryAfterSeconds { get; }

    /// <summary>
    /// The rate limit category that was exceeded.
    /// </summary>
    public string? Category { get; }

    /// <summary>
    /// The current limit value for this category.
    /// </summary>
    public long? Limit { get; }

    /// <summary>
    /// The number of requests remaining in the current window.
    /// </summary>
    public long? Remaining { get; }

    public AinosRateLimitException()
    {
    }

    public AinosRateLimitException(string message, int retryAfterSeconds)
        : base(message)
    {
        RetryAfterSeconds = retryAfterSeconds;
    }

    public AinosRateLimitException(string message, int retryAfterSeconds, string category, long limit, long remaining)
        : base($"Rate limit exceeded for '{category}': {message}. Retry after {retryAfterSeconds}s. ({remaining}/{limit} remaining)")
    {
        RetryAfterSeconds = retryAfterSeconds;
        Category = category;
        Limit = limit;
        Remaining = remaining;
    }
}